File: lab-1/src/main/java/com/evist0/labforwardlist/MainController.java
package com.evist0.labforwardlist;

import com.evist0.labforwardlist.factory.ObjectFactory;
import com.evist0.labforwardlist.factory.ObjectFactoryRegistry;
import com.evist0.labforwardlist.model.ForwardList;
import javafx.beans.property.Property;
import javafx.beans.property.SimpleStringProperty;
import javafx.collections.FXCollections;
import javafx.collections.ObservableList;
import javafx.fxml.FXML;
import javafx.fxml.Initializable;
import javafx.scene.control.Button;
import javafx.scene.control.ChoiceBox;
import javafx.scene.control.ListView;
import javafx.scene.control.TextField;

import java.io.*;
import java.net.URL;
import java.util.Collections;
import java.util.ResourceBundle;

public class MainController implements Initializable {
    private static final String DUMP_FILENAME = "save.bin";

    @FXML
    private ChoiceBox<String> typeChoiceBox;
    @FXML
    private TextField inputTextField;
    @FXML
    private Button addButton;
    @FXML
    private Button insertButton;
    @FXML
    private Button removeButton;
    @FXML
    private Button sortButton;
    @FXML
    private Button loadButton;
    @FXML
    private Button saveButton;
    @FXML
    private ListView itemListView;

    private final ObservableList<String> typeObservableList;
    private final Property<String> selectedTypeProperty;
    private ForwardList forwardList;

    public MainController() {
        typeObservableList = FXCollections.observableList(ObjectFactoryRegistry.getAllFactories());
        selectedTypeProperty = new SimpleStringProperty(typeObservableList.get(0));
    }

    @Override
    public void initialize(URL url, ResourceBundle resourceBundle) {
        typeChoiceBox.setItems(typeObservableList);
        typeChoiceBox.valueProperty().bindBidirectional(selectedTypeProperty);

        selectedTypeProperty.addListener((observableValue, old, value) -> {
            forwardList = null;
            onItemCollectionChanged();
        });

        addButton.setOnAction(actionEvent -> {
            if (forwardList == null) {
                forwardList = new ForwardList();
            }

            forwardList.add(getObjectFactory().parse(inputTextField.getText()));
            onItemCollectionChanged();
        });

        insertButton.setOnAction(actionEvent -> {
            int at = itemListView.getSelectionModel().getSelectedIndex();

            forwardList.insert(at, getObjectFactory().parse(inputTextField.getText()));
            onItemCollectionChanged();
        });

        removeButton.setOnAction(actionEvent -> {
            int at = itemListView.getSelectionModel().getSelectedIndex();

            forwardList.remove(at);
            onItemCollectionChanged();
        });

        sortButton.setOnAction(actionEvent -> {
            forwardList.sort(getObjectFactory().getComparator());
            onItemCollectionChanged();
        });

        loadButton.setOnAction(actionEvent -> {
            try (FileInputStream fileInputStream = new FileInputStream(DUMP_FILENAME)) {
                ObjectInputStream objectInputStream = new ObjectInputStream(fileInputStream);

                String factoryName = (String) objectInputStream.readObject();
                selectedTypeProperty.setValue(factoryName);

                forwardList = new ForwardList();

                boolean eof = false;
                while (!eof) {
                    try {
                        Object readObject = objectInputStream.readObject();
                        forwardList.add(readObject);
                    } catch (EOFException exception) {
                        eof = true;
                    }
                }

                onItemCollectionChanged();
            } catch (IOException | ClassNotFoundException e) {
                throw new RuntimeException(e);
            }
        });

        saveButton.setOnAction(actionEvent -> {
            try (FileOutputStream fileOutputStream = new FileOutputStream(DUMP_FILENAME)) {
                ObjectOutputStream objectOutputStream = new ObjectOutputStream(fileOutputStream);

                String factoryName = getObjectFactory().getObjectName();
                objectOutputStream.writeObject(factoryName);

                for (Object i : forwardList) {
                    objectOutputStream.writeObject(i);
                }
            } catch (IOException e) {
                throw new RuntimeException(e);
            }
        });
    }

    private ObjectFactory getObjectFactory() {
        return ObjectFactoryRegistry.getFactory(selectedTypeProperty.getValue());
    }

    private void onItemCollectionChanged() {
        if (forwardList == null) {
            itemListView.setItems(FXCollections.observableList(Collections.emptyList()));
            return;
        }

        itemListView.setItems(FXCollections.observableList(forwardList.toList()));
    }
}

File: lab-2/src/main/scala/com/hamaddo/labforwardlist/MainController.scala
package com.hamaddo.labforwardlist

import com.hamaddo.labforwardlist.factory.{ObjectFactory, ObjectFactoryRegistry}
import com.hamaddo.labforwardlist.model.ForwardList
import javafx.beans.property.SimpleStringProperty
import javafx.collections.{FXCollections, ObservableList}
import scalafx.scene.control.{Button, ChoiceBox, ListView, TextField}
import scala.jdk.CollectionConverters._
import scalafxml.core.macros.sfxml
import java.io.{EOFException, FileInputStream, FileOutputStream, IOException, ObjectInputStream, ObjectOutputStream}
import java.util.Collections

@sfxml
class MainController(
                      private val typeChoiceBox: ChoiceBox[String],
                      private val inputTextField: TextField,
                      private val addButton: Button,
                      private val insertButton: Button,
                      private val removeButton: Button,
                      private val sortButton: Button,
                      private val loadButton: Button,
                      private val saveButton: Button,
                      private val itemListView: ListView[Object]) {
  private val DUMP_FILENAME = "save.bin"

  private val typeObservableList: ObservableList[String] = FXCollections.observableList(ObjectFactoryRegistry.getAllFactories)
  private val selectedTypeProperty = new SimpleStringProperty(typeObservableList.get(0))

  private var forwardList = new ForwardList

  typeChoiceBox.setItems(typeObservableList)
  typeChoiceBox.valueProperty().bindBidirectional(selectedTypeProperty)

  selectedTypeProperty.addListener((_, _, _) => {
    forwardList = null
  })

  addButton.onAction = _ => {
    if (forwardList == null) {
      forwardList = new ForwardList
    }

    val value = inputTextField.getText
    val parsedValue = getObjectFactory.parse(value)

    forwardList.add(parsedValue)

    onItemCollectionChanged()
  }

  insertButton.onAction = _ => {
    val at = itemListView.getSelectionModel.getSelectedIndex

    forwardList.insert(at, getObjectFactory.parse(inputTextField.getText))
    onItemCollectionChanged()
  }

  removeButton.onAction = _ => {
    val at = itemListView.getSelectionModel.getSelectedIndex

    forwardList.remove(at)
    onItemCollectionChanged()
  }

  sortButton.onAction = _ => {
    forwardList.sort(getObjectFactory.getComparator)
    onItemCollectionChanged()
  }

  loadButton.onAction = _ => {
    try {
      val fileInputStream = new FileInputStream(DUMP_FILENAME)

      try {
        val objectInputStream = new ObjectInputStream(fileInputStream)

        val factoryName = objectInputStream.readObject.asInstanceOf[String]
        selectedTypeProperty.setValue(factoryName)

        forwardList = new ForwardList

        var eof = false
        while (!eof) try {
          val readObject = objectInputStream.readObject
          forwardList.add(readObject)
        } catch {
          case _: EOFException =>
            eof = true
        }

        onItemCollectionChanged()
      } catch {
        case e@(_: IOException | _: ClassNotFoundException) =>
          throw new RuntimeException(e)
      } finally {
        if (fileInputStream != null) {
          fileInputStream.close()
        }
      }
    }
  }

  saveButton.onAction = _ => {
    try {
      val fileOutputStream = new FileOutputStream(DUMP_FILENAME)

      try {
        val objectOutputStream = new ObjectOutputStream(fileOutputStream)

        val factoryName = getObjectFactory.getObjectName
        objectOutputStream.writeObject(factoryName)

        for (i <- forwardList) {
          objectOutputStream.writeObject(i)
        }
      } catch {
        case e: IOException =>
          throw new RuntimeException(e)
      } finally {
        if (fileOutputStream != null) {
          fileOutputStream.close()
        }
      }
    }
  }

  private def getObjectFactory: ObjectFactory = {
    ObjectFactoryRegistry.getFactory(selectedTypeProperty.getValue)
  }

  private def onItemCollectionChanged(): Unit = {
    if (forwardList == null) {
      itemListView.setItems(FXCollections.observableList(Collections.emptyList[Object]))
      return
    }

    itemListView.setItems(FXCollections.observableList(forwardList.toList.asJava))
  }
}
